hardware_interface/include/hardware_interface/actuator_interface.hpp

```diff
@@ -115,33 +120,202 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
    *
    * Note the ownership over the state interfaces is transferred to the caller.
    *
    * \return vector of state interfaces
    */
-  virtual std::vector<StateInterface> export_state_interfaces() = 0;
+  [[deprecated(
+    "Replaced by vector<std::shared_ptr<StateInterface>> on_export_state_interfaces() method. "
+    "Exporting is "
+    "handled "
+    "by the Framework.")]] virtual std::vector<StateInterface>
+  export_state_interfaces()
+  {
+    // return empty vector by default. For backward compatibility we try calling
+    // export_state_interfaces() and only when empty vector is returned call
+    // on_export_state_interfaces()
+    return {};
+  }
+
+  /**
+   * Override this method to export custom StateInterfaces which are not defined in the URDF file.
+   * Those interfaces will be added to the unlisted_state_interfaces_ map.
+   *
+   *  Note method name is going to be changed to export_state_interfaces() as soon as the deprecated
+   * version is removed.
+   *
+   * \return vector of descriptions to the unlisted StateInterfaces
+   */
+  virtual std::vector<hardware_interface::InterfaceDescription>
+  export_state_interface_descriptions()
+  {
+    // return empty vector by default.
+    return {};
+  }
+
+  /**
+   * Default implementation for exporting the StateInterfaces. The StateInterfaces are created
+   * according to the InterfaceDescription. The memory accessed by the controllers and hardware is
+   * assigned here and resides in the actuator_interface.
+   *
+   * \return vector of shared pointers to the created and stored StateInterfaces
+   */
+  virtual std::vector<std::shared_ptr<StateInterface>> on_export_state_interfaces()
+  {
+    // import the unlisted interfaces
+    std::vector<hardware_interface::InterfaceDescription> unlisted_interface_descriptions =
+      export_state_interface_descriptions();
+
+    std::vector<std::shared_ptr<StateInterface>> state_interfaces;
+    state_interfaces.reserve(
+      unlisted_interface_descriptions.size() + joint_state_interfaces_.size());
+
+    // add InterfaceDescriptions and create the StateInterfaces from the descriptions and add to
+    // maps.
+    for (const auto & description : unlisted_interface_descriptions)
+    {
+      auto name = description.get_name();
+      unlisted_state_interfaces_.insert(std::make_pair(name, description));
+      auto state_interface = std::make_shared<StateInterface>(description);
+      actuator_states_.insert(std::make_pair(name, state_interface));
+      unlisted_states_.push_back(state_interface);
+      state_interfaces.push_back(state_interface);
+    }
+
+    for (const auto & [name, descr] : joint_state_interfaces_)
+    {
+      auto state_interface = std::make_shared<StateInterface>(descr);
+      actuator_states_.insert(std::make_pair(name, state_interface));
+      joint_states_.push_back(state_interface);
+      state_interfaces.push_back(state_interface);
+    }
+    return state_interfaces;
+  }
```

Comment:
Maybe we can leave this to the user by providing some helper methods, so he can define it himself in the same structure or a different one. Even he will be open to do it in an unordered_map, map, or a vector

Same comments for command interfaces as well

Reply:
Sure 👍🏾

Reply:
I agree with Bence on this.

Reply:
Sure! I'm fine with this